### PR TITLE
Improve screen api performance

### PR DESCRIPTION
**Pull request by ryanmelt**

### CODE REVIEW

File: openc3/lib/openc3/api/tlm_api.rb [hunk 1]
```diff
@@ -227,16 +227,16 @@ def get_tlm_buffer(target_name, packet_name, scope: $openc3_scope, token: $openc
...
       packet_name = packet_name.upcase
       authorize(permission: 'tlm', target_name: target_name, packet_name: packet_name, scope: scope, token: token)
       packet = TargetModel.packet(target_name, packet_name, scope: scope)
       t = _validate_tlm_type(type)
       raise ArgumentError, "Unknown type '#{type}' for #{target_name} #{packet_name}" if t.nil?
       items = packet['items'].map { | item | item['name'].upcase }
-      cvt_items = items.map { | item | "#{target_name}__#{packet_name}__#{item}__#{type}" }
-      current_values = CvtModel.get_tlm_values(cvt_items, stale_time: stale_time, scope: scope)
+      cvt_items = items.map { | item | [target_name, packet_name, item, type] }
+      current_values = CvtModel.get_tlm_values(cvt_items, stale_time: stale_time, cache_timeout: cache_timeout, scope: scope)
```
Comment (jmthomas) [inline openc3/lib/openc3/api/tlm_api.rb:239]: Are you changing this method to take an array of array of strings? It used to be an array of strings with the underscores.

**Review by jmthomas — APPROVED** [review]
Looks like this adds a really quick in-memory cache for simultaneous telemetry lookups resulting from screens calling get_tlm_values. Nice!

File: openc3/lib/openc3/models/auth_model.rb [hunk 2]
```diff
@@ -28,6 +28,12 @@ class AuthModel
     PRIMARY_KEY = 'OPENC3__TOKEN'
     SERVICE_KEY = 'OPENC3__SERVICE__TOKEN'
 
+    TOKEN_CACHE_TIMEOUT = 5
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/models/auth_model.rb:31]: This is how long an old password will still work.  No affect on Enterprise, only open source.

File: openc3/lib/openc3/api/tlm_api.rb [hunk 3]
```diff
@@ -66,30 +66,30 @@ module Api
     # @param args [String|Array<String>] See the description for calling style
     # @param type [Symbol] Telemetry type, :RAW, :CONVERTED (default), :FORMATTED, or :WITH_UNITS
     # @return [Object] The telemetry value formatted as requested
-    def tlm(*args, type: :CONVERTED, scope: $openc3_scope, token: $openc3_token)
-      target_name, packet_name, item_name = tlm_process_args(args, 'tlm', scope: scope)
+    def tlm(*args, type: :CONVERTED, cache_timeout: 0.1, scope: $openc3_scope, token: $openc3_token)
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/api/tlm_api.rb:69]: That's why I chose that as default.  It still supports 10Hz, but any faster queries will use the cache.

File: openc3/lib/openc3/models/cvt_model.rb [hunk 4]
```diff
@@ -111,18 +109,19 @@ def self.get_item(target_name, packet_name, item_name, type:, scope: $openc3_sco
...
+      now = Time.now
       results = []
       lookups = []
       packet_lookup = {}
       overrides = {}
       # First generate a lookup hash of all the items represented so we can query the CVT
-      items.each { |item| _parse_item(lookups, overrides, item, scope: scope) }
+      items.each { |item| _parse_item(now, lookups, overrides, item, cache_timeout: cache_timeout, scope: scope) }
 
+      now = now.to_f
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/models/cvt_model.rb:121]: Its used at line 137

File: openc3/lib/openc3/models/cvt_model.rb [hunk 5]
```diff
@@ -232,23 +235,88 @@ def self.normalize(target_name, packet_name, item_name, type: :ALL, scope: $open
...
+      tgt_pkt_key = "#{scope}__tlm__#{target_name}__#{packet_name}"
+      @@override_cache[tgt_pkt_key] = [Time.now, hash]
       if hash.empty?
         Store.hdel("#{scope}__override__#{target_name}", packet_name)
       else
         Store.hset("#{scope}__override__#{target_name}", packet_name, JSON.generate(hash.as_json(:allow_nan => true)))
       end
     end
 
+    def self.determine_latest_packet_for_item(target_name, item_name, cache_timeout: 0.1, scope: $openc3_scope)
```
[the remark this answers is not in the document]
Comment (ryanmelt) [reply openc3/lib/openc3/models/cvt_model.rb:248]: Its way faster.  LATEST was not optimized at all and could requires 100s of Redis queries.  Now it takes only the exact number of packets that contain the item.

File: openc3/lib/openc3/api/tlm_api.rb (hunk 1 above)
Comment (ryanmelt) [reply openc3/lib/openc3/api/tlm_api.rb:239]: Yes, this was just an optimization to prevent splitting twice.

Comment (jmthomas) [reply openc3/lib/openc3/api/tlm_api.rb:239]: Won't that break existing code?

Comment (ryanmelt) [reply openc3/lib/openc3/api/tlm_api.rb:239]: If anyone is using CvtModel.get_tlm_values this is a breaking change.  That's an internal api though.

**Pull request merged**